threescale/src/utils.rs

```diff
@@ -1,23 +1,54 @@
-use crate::structs::Period;
-use crate::structs::{Application, ThreescaleData};
-use threescalers::response::Period as ResponsePeriod;
+use crate::structs::{Application, Period, ThreescaleData};
+use std::time::Duration;
 
-// Perform metrics update based on threescale specific logic
-pub fn update_metrics(_new_hits: &ThreescaleData, _application: &mut Application) -> bool {
-    true
+#[derive(Debug, Clone, thiserror::Error)]
+pub enum UpdateMetricsError {
+    #[error("overflow due to two duration addition")]
+    DurationOverflow,
+    #[error("request is rate-limited")]
+    RateLimited,
 }
 
-impl From<&ResponsePeriod> for Period {
-    fn from(res_period: &ResponsePeriod) -> Self {
-        match res_period {
-            ResponsePeriod::Minute => Period::Minute,
-            ResponsePeriod::Hour => Period::Hour,
-            ResponsePeriod::Day => Period::Day,
-            ResponsePeriod::Week => Period::Week,
-            ResponsePeriod::Month => Period::Month,
-            ResponsePeriod::Year => Period::Year,
-            ResponsePeriod::Eternity => Period::Eternity,
-            _ => Period::Eternity,
+// updates application to reflect consumed quota if not rate-limited
+// returns Ok() if not rate-limited and faced no problem updating the application
+pub fn limit_check_and_update_application(
+    data: &ThreescaleData,
+    app: &mut Application,
+    current_time: &Duration,
+) -> Result<(), UpdateMetricsError> {
+    for (metric, hits) in data.metrics.borrow().iter() {
+        // note: we assume missing metrics are not limited until new state is fetched
+        if let Some(usage_report) = app.local_state.get_mut(metric) {
+            let mut period = &mut usage_report.period_window;
+
+            // taking care of period window expiration
+            if period.window != Period::Eternity && period.end < *current_time {
+                let time_diff = current_time
+                    .checked_sub(period.start)
+                    .ok_or(UpdateMetricsError::DurationOverflow)?;
+                let num_windows = time_diff.as_secs() / period.window.as_secs();
+                let seconds_to_add = num_windows * period.window.as_secs();
+
+                // set to new period window
+                period.start = period
+                    .start
+                    .checked_add(Duration::from_secs(seconds_to_add))
+                    .ok_or(UpdateMetricsError::DurationOverflow)?;
+
+                period.end = period
+                    .end
+                    .checked_add(Duration::from_secs(seconds_to_add))
+                    .ok_or(UpdateMetricsError::DurationOverflow)?;
+
+                // reset left hits back to max value
+                usage_report.left_hits = usage_report.max_value;
+            }
+
+            if usage_report.left_hits < *hits {
+                return Err(UpdateMetricsError::RateLimited);
+            }
+            usage_report.left_hits -= *hits;
```

Comment:
I'm not very sure about updating the usages this way. I thought we'll need a more complex logic with the hierarchies as well. I went through the section that the auth-filter([https://github.com/3scale/threescale-wasm-auth/blob/4dfcf9ae9e9b6d9651880c3c38eef8082ef6f0f5/src/proxy/authrep.rs#L90()) sets usages and seems like we can straightaway update the usages like that since auth-filter maps the path with usage rules and sends us only the metrics related to this particular url. But I'm not sure whether we need to handle scenarios like `update parent only` or `update child only`. I remember something like that from previous conversations.

Reply:
My bad! I forgot to implement `add_hierarchy_to_metrics` before pushing that code; It's now up and you can review it.